app/src/test/java/ch/hikemate/app/model/route/HikesViewModelTest.kt

```diff
@@ -385,29 +400,31 @@ class HikesViewModelTest {
         assertEquals(singleOsmHike1[0].id, hikesViewModel.hikeFlows.value[0].value.id)
         // Check that the loaded hike is now marked as saved
         assertTrue(hikesViewModel.hikeFlows.value[0].value.isSaved)
+        // The loading error message should still be null
+        assertNull(hikesViewModel.loadingErrorMessageId.value)
       }
 
   @Test
   fun `refreshSavedHikesCache sets loading to true then false`() =
       runTest(dispatcher) {
-        // Listen to the changes made to loading during the call
-        val emissions = mutableListOf<Boolean>()
-        val job = backgroundScope.launch { hikesViewModel.loading.collect { emissions.add(it) } }
+        // Before launching the operation, loading should be false
+        assertFalse(hikesViewModel.loading.value)
 
         // Set the repository to throw an exception because we do not care
-        coEvery { savedHikesRepo.loadSavedHikes() } throws Exception("Failed to load saved hikes")
+        coEvery { savedHikesRepo.loadSavedHikes() } answers
```

Comment:
Thank you for your warning, although I'm not sure how exactly this case could happen. Were you able to observe it?

`runTest` is used with a `dispatcher` parameter. The dispatcher is an `UnconfinedTestDispatcher`, meaning it runs all the coroutines started on it pretty much instantaneously. The test shouldn't end before all of its coroutines are done. The same dispatcher is provided to the `HikesViewModel` instance and used as main dispatcher, using

```kotlin
Dispatchers.setMain(dispatcher)
```

So basically, as far as I understand, everyone uses the same scheduler and all coroutines should terminate before the end of the test.

Perhaps you could point me to the fix the site proposes to solve this? I'm not sure what I should look for on the link you provided.

Reply:
So a way you can test it is by:

1. Delete those lines in `HikesViewModel.refreshSavedHikesCache`:
```kotlin
// Let the user know a heavy load operation is being performed
setLoading(value = true, errorMessageId = null)
```
2. Run the test `refreshSavedHikesCache sets loading to true then false`

The test should fail, because loading is not set to true in the function.

Reply:
But you are right, I should add something that checks whether the coroutine has terminated.

Reply:
I just realized, I actually already do that, on this line :

```kotlin
// Check that the operation occurred, i.e. the repo was called
coVerify(exactly = 1) { savedHikesRepo.loadSavedHikes() }
```

This effectively checks that the function was called and terminated, doesn't it? Maybe it doesn't check that it terminated, but I'm not sure I need to add an additional check. What do you think?